sycl/include/sycl/ext/intel/esimd/memory.hpp

```diff
@@ -236,6 +254,23 @@ __ESIMD_API void scatter(Tx *p, simd_view<Toffset, RegionTy> offsets,
   scatter<Tx, N>(p, simd<Ty, N>(offsets), vals, mask);
 }
 
+/// A variation of \c scatter API with \c offsets represented as scalar.
+///
+/// @tparam Tx Element type, must be of size 4 or less.
+/// @tparam N Number of elements to write; can be \c 1, \c 2, \c 4, \c 8, \c 16
+///   or \c 32.
+/// @param p The base address.
+/// @param offset the scalar 32-bit or 64-bit offset in bytes.
+/// ((byte*)p + offset) must be element size aligned.
+/// @param vals The vector to scatter.
+/// @param mask The access mask, defaults to all 1s.
+///
+template <typename Tx, int N, typename Toffset>
+__ESIMD_API std::enable_if_t<std::is_integral_v<Toffset>>
+scatter(Tx *p, Toffset offset, simd<Tx, N> vals, simd_mask<N> mask = 1) {
+  scatter<Tx, N>(p, simd<Toffset, N>(offset), vals, mask);
```

Comment:
This operation is Ok only when N==1.
If N > 1, then it is basically undefined-behavior.

 I suggest allowing scalar offset for STORE semantic functions (i.e. for scatter/store and atomic_op::store) only if N==1.

Reply:
1) If you talk about binomial* test, then this test is owned by ESIMD team. 
It uses scalar offset for gather only, which is in my opinion is inefficient as scalar_load + broadcast generates better code. It has well defined behavior for loads.
For scatters the test uses vector 'simd<uint32_t, SIMD_WIDTH> offsets;'. 

2) I agree we do not have any control on situations like scatter<int, 16>(p, simd<int, 16>(5), vals, mask). 
Doing so is wrong, it is solely user's responsibility, and code doing that should be fixed.
My point is that adding the new scatter/store prototypes you added in this PR provokes users to use SCALAR offset for (N > 1), which is always undefined behavior, and should be prohibited at compile time.

I still suggest allowing only N==1 for scatter and atomic_update(atomic_op::store).

Reply:
If 'simd<uint32_t, SIMD_WIDTH> offsets' is initialized with a const, then the test is bad and needs some fix. The one who wrote it was focused on performance aspect of the test and missed a) one of performance aspects because scalar store is faster than saving N values to the same address anb b) correctness aspect - doing scatter of N values to to 1 scalar address is UB in general.

Regarding N as a template parameter - actually N is a template parameter.
The code now is:
```
template <typename Tx, int N, typename Toffset>
__ESIMD_API std::enable_if_t<std::is_integral_v<Toffset>>
scatter(Tx *p, Toffset offset, simd<Tx, N> vals, simd_mask<N> mask = 1) {
  scatter<Tx, N>(p, simd<Toffset, N>(offset), vals, mask);
```
It can be changed to:
```
template <typename Tx, int N, typename Toffset>
__ESIMD_API std::enable_if_t<std::is_integral_v<Toffset> && N == 1>
scatter(Tx *p, Toffset offset, simd<Tx, N> vals, simd_mask<N> mask = 1) {
  scatter<Tx, N>(p, simd<Toffset, N>(offset), vals, mask);
```